notanorm/jsondb.py

```diff
@@ -0,0 +1,571 @@
+
+import logging
+
+log = logging.getLogger(__name__)
+
+
+class QueryRes:
+    rowcount = 0
+    lastrowid = 0
+
+    def __init__(self, parent: "JsonDb", *, generator=None):
+        self.parent = parent
+        self.generator = generator
+
+    def execute(self, sql, parameters=()):
+        self.parent._executeone(self, sql, parameters)
+
+    def fetchall(self):
+        ret = []
+        if self.generator:
+            for ent in self.generator:
+                ret.append(ent)
+        return ret
+
+    def fetchone(self):
+        if self.generator:
+            try:
+                return next(self.generator)
+            except StopIteration:
+                return None
+
+    def close(self):
+        self.generator and self.generator.close()
+
+
+class JsonDb(DbBase):
+    uri_name = "jsondb"
+    use_pooled_locks = False
+    generator_guard = False
+    max_reconnect_attempts = 1
+    retry_file_access = 5
+    _parse_memo = {}
+    default_values = " () values ()"
+    closed = False
+
+    def _begin(self, conn: "JsonDb"):
+        self._tx.append(copy.deepcopy(self.__dat))
+
+    def commit(self):
+        self._tx = self._tx[0:-1]
+        if not self._tx:
+            self.__write()
+
+    def serialize(self, val):
+        if isinstance(val, (int, str, float, bool, type(None))):
+            return val
+        tname = type(val).__name__
+        return {"type": tname, "value": getattr(self, "serialize_" + tname)(val)}
+
+    def deserialize(self, obj):
+        return getattr(self, "deserialize_" + obj["type"])(obj["value"])
+
+    def serialize_bytes(self, val):
+        return base64.b64encode(val).decode("utf8")
+
+    def deserialize_bytes(self, val):
+        return base64.b64decode(val)
+
+    def __write(self):
+        if not self.__is_mem:
+            # tmp file is pid + thread, so they don't accumulate forever on crashes
+            tmp = (
+                self.__file
+                + ".jtmp."
+                + str(self.__pid)
+                + "."
+                + str(threading.get_ident())
+            )
+            dat = copy.deepcopy(self.__dat)
+            for tab, rows in dat.items():
+                for i, row in enumerate(rows):
+                    for col, val in row.items():
+                        dat[tab][i][col] = self.serialize(val)
+            try:
+                with self.__retry_fileop(lambda: open(tmp, "w")) as fp:
+                    # not safe to retry this!
+                    json.dump(dat, fp)
+                self.__retry_fileop(lambda: os.replace(tmp, self.__file))
+            except Exception as ex:
+                with contextlib.suppress(Exception):
+                    os.unlink(tmp)
+                raise ex
+        self.__dirty = False
+
+    def __retry_fileop(self, func: Callable) -> Any:
+        last_ex = None
+        for _ in range(self.retry_file_access):
+            try:
+                return func()
+            except PermissionError as ex:
```

Comment:
in real life tests on windows with many processes, i encountered frequent PermissionError errors when multiple writers were writing.   i know that cvfs doesn't care about this use case (we have a single writer), but since it was a observed, i thought it was good to handle it